aeppl/scan.py

```diff
@@ -0,0 +1,332 @@
+
+from aeppl.abstract import MeasurableVariable
+from aeppl.logprob import _logprob
+from aeppl.opt import logprob_rewrites_db, rv_sinking_db
+
+
+class MeasurableScan(Scan):
+    """A placeholder used to specify a log-likelihood for a scan sub-graph."""
+
+    @classmethod
+    def create_node(cls, node, rv_outs):
+        mscan_op = copy(node.op)
+        mscan_op.__class__ = cls
+        new_outputs = [o.clone() for o in node.outputs]
+        new_node = Apply(mscan_op, node.inputs, new_outputs)
+        mscan_op.scan_args = ScanArgs.from_node(new_node)
+        return new_node
+
+
+MeasurableVariable.register(MeasurableScan)
+
+
+@local_optimizer([Scan])
+def push_out_rvs_from_scan(fgraph, node):
+    r"""Push `MeasurableVariable`\s out of `Scan` nodes.
+
+    When `MeasurableVariable`\s are created within the inner-graph of a `Scan` and
+    are not output to the outer-graph, we "push" them out of the inner-graph.
+    This helps us produce an outer-graph in which all the relevant
+    `MeasurableVariable`\s are accessible (e.g. for constructing a log-likelihood
+    graph).
+    """
+
+    if not isinstance(node.op, Scan):
+        return None
+
+    if isinstance(node.op, MeasurableScan):
+        return
+
+    curr_scanargs = ScanArgs(
+        node.inputs, node.outputs, node.op.inputs, node.op.outputs, node.op.info
+    )
+
+    # Find the un-output `MeasurableVariable`s created in the inner-graph
+    clients: Dict[Variable, List[Variable]] = {}
+
+    local_fgraph_topo = aesara.graph.basic.io_toposort(
+        curr_scanargs.inner_inputs, curr_scanargs.inner_outputs, clients=clients
+    )
+    unpushed_inner_rvs = []
+    for n in local_fgraph_topo:
+        if isinstance(n.op, MeasurableVariable):
+            unpushed_inner_rvs.extend(
+                [c for c in clients[n] if c not in curr_scanargs.inner_outputs]
+            )
+
+    if len(unpushed_inner_rvs) != 0:
+        # This means that user used function like
+        # def scan_fn(*args):
+        #     S_t = at.random.categorical()
+        #     Y_t = at.random.normal(arg1[S_t])
+        # return Y_t
+        raise ValueError(
+            f"The Random Variables {unpushed_inner_rvs} were"
+            "not returned as outputs by Scan's inner function"
+        )
+
+    op = MeasurableScan(
+        curr_scanargs.inner_inputs, curr_scanargs.inner_outputs, curr_scanargs.info
+    )
+    new_node = op.make_node(*curr_scanargs.outer_inputs)
+    op.scan_args = ScanArgs.from_node(new_node)
+
+    outputs = list(new_node.outputs)
+
+    # Return only the replacements for the original `node.outputs`
+    new_inner_out_idx = [
+        curr_scanargs.inner_outputs.index(i) for i in unpushed_inner_rvs
+    ]
+    _ = [
+        outputs.pop(op.var_mappings["outer_out_from_inner_out"][i])
+        for i in new_inner_out_idx
+    ]
+
+    # TODO: Remap any value variables associated with the old `Scan` to this
+    # new `MeasurableScan`?
```

Comment:
I think the `rv_value_remapper` in `joint_logprob` takes care of that no ?

The only issue is it doesn't map variables that are not the `var` (i.e. `S_rv` in our test cases) is that what you meant over here.

Reply:
> I think the `rv_remapper` in `joint_logprob` takes care of that no ?

Yes, it generally should.

> The only issue is it doesn't map variables that are not the `var` (i.e. `S_rv` in our test cases). Is that what you meant over here ?

The `PreserveRVMappings` feature might not work for variables that aren't the literal inputs/outputs of a replacement.  I believe that was one of my concerns.  That might not be an issue, though.  We really need to set up some example and work through them to find out.

Reply:
It looks like [the current issue](https://github.com/aesara-devs/aeppl/pull/24/checks?check_run_id=3357658699#step:6:256) actually _is_ due to `PreserveRVMappings` not handling every variable that was replaced.

In this particular case, `PreserveRVMappings.on_change_input` is only ever called with `r` (i.e. the variable being replaced) equal to `Y_rv`.  This is because the `FunctionGraph` that's being used has only one output: `Y_rv`.  The second output of the `Scan` is `S_rv` and the test fails when that output appears in the `joint_logprob` loop.

If `PreserveRVMappings.on_change_input` were called with `r` equal to `S_rv` and `new_r` equal to the corresponding new output produced by `push_out_rvs_from_scan`, this problem wouldn't appear.

We might be able to add all the random variables as outputs to the `fgraph` created in `joint_logprob` and get this to work.  That would guarantee that `PreserveRVMappings.on_change_input` would be called for each one.

Reply:
I've just added a fix to that effect.